Sprint 10: controllers-films-users.

Novacancy531 opened the pull request:
Реализация задания 10 спринта.

Novac531 committed:
Development:
added UserController,
added FilmController,
added GlobalExceptionHandler,
added custom check AfterDate,
added custom exception ConditionsNotMetException,
added SLF4J logger,
added JUnit tests for UserController & FilmController,
changed README.md

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,60 @@
+    Map<Long, Film> films = new HashMap<>();
+
+    @PostMapping
+    public Film post(@Valid @RequestBody Film film) {
```
Методы контроллера лучше называть не по имени http-метода, а по действию, которое совершается.
Например createFilm().

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,60 @@
+@Validated
+@RequestMapping("/films")
 public class FilmController {
+    Map<Long, Film> films = new HashMap<>();
```
модификатор доступа?

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,7 +1,60 @@
+    }
+
+    private long getNextId() {
+        long currentMaxId = films.keySet()
```
То что пользуешься stream это хорошо, но сохранение счетчика в поле будет производительней, т.к. не нужно будет перебирать весь массив фильмов. В текущей реализации чем больше фильмов, тем медленней будет работать метод getNextId().

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/annotation/AfterDate.java:
```diff
@@ -0,0 +1,17 @@
+@Target({ElementType.FIELD})
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = AfterDateValidator.class)
+public @interface AfterDate {
```
Отлично, что реализовал сваю аннотацию.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -0,0 +1,64 @@
+@RestController
+@Validated
+@RequestMapping("/users")
+public class UserController {
```
Те же замечания, что и для фильмов.

VadimZharkov commented on src/main/java/ru/yandex/practicum/filmorate/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,56 @@
+        return ResponseEntity.status(status).body(error);
+    }
+
+        public record ErrorResponse(LocalDateTime timestamp, int status, String error, String message) {
```
Отлично, что использовал record

VadimZharkov reviewed (changes requested):
Хорошая работа, можно улучшить.

Novacancy531 committed:
Исправления:
1. Для мап user и film добавлен модификатор доступа.
2. Заменен метод присвоения id.
3. Заменены названия методов контроллеров.
